examples/wasm/envoy_filter_http_wasm_example.cc

```diff
@@ -1,14 +1,219 @@
+  }
+};
+
+/*
+ * Check if the string val exists in the vector vec
+ */
+inline bool exists(std::string val, std::vector<std::string> vec) {
+  return std::find(vec.begin(), vec.end(), val) != vec.end();
+}
+
+/*
+ * Validate and store a field in Config
+ * Input:
+ *   field: a json object to be parsed, either query param, header, or cookie
+ *   include: a pointer to store the parsed result (<field_include in Config)
+ *   params: a pointer to store the parsed result (<field>s in Config)
+ * Output:
+ *   true on success
+ *   false on failure (if both 'include' and 'exclude' are present in the field)
+ */
+bool validate_config_field(json field, bool* include, std::vector<std::string>* params) {
+   if (field.is_null()) {
+     return true;
+   }
+   if (!field["include"].is_null() && !field["exclude"].is_null()) {
+     LOG_ERROR("onConfigure: \"include\" and \"exclude\" should not be used simultaneously");
+     return false;
+   }
+   if (!field["include"].is_null()) {
+     *include = true;
+     auto include_params = field["include"].get<std::vector<std::string>>();
+     params->insert(params->end(), include_params.begin(), include_params.end());
+   }
+   if (!field["exclude"].is_null()) {
+     *include = false;
+     *params = field["exclude"].get<std::vector<std::string>>();
+   }
+   return true;
+}
+
+/*
+ * Methods for parsing query parameters
+ */
+QueryParams parseParameters(std::string data, size_t start,
+                            std::string delim = "&", std::string eq = "=") {
+  QueryParams params;
+
+  while (start < data.size()) {
+    size_t end = data.find(delim, start);
+    if (end == std::string::npos) {
+      end = data.size();
+    }
+    std::string param = data.substr(start, end - start);
+
+    const size_t equal = param.find(eq);
+    if (equal != std::string::npos) {
+      params.emplace(param.substr(start, start + equal),
+                     param.substr(start + equal + 1, end));
+    } else {
+      params.emplace(param.substr(start, end), "");
+    }
+
+    start = end + 1;
+  }
+
+  return params;
+}
+
+QueryParams parsePath(std::string path) {
+  size_t start = path.find('?');
+  if (start == std::string::npos) {
+    QueryParams params;
+    return params;
+  }
+  start++;
+  return parseParameters(path, start);
+}
+
+QueryParams parseBody(std::string body) {
+  return parseParameters(body, 0);
+}
+
+QueryParams parseCookie(std::string cookie) {
+  return parseParameters(cookie, 0, "; ");
+}
+
+std::string toString(QueryParams params) {
+  std::string str;
+  for (auto param : params) {
+    str += param.first + " -> " + param.second;
+  }
+  return str;
+}
+
+/*
+ * Detect SQL injection on given input string
+ */
+bool detectSQLi(std::string input, std::string key, std::string part) {
+  struct libinjection_sqli_state state;
+  char* input_char_str = const_cast<char*>(input.c_str());
+  libinjection_sqli_init(&state, input_char_str, input.length(), FLAG_NONE);
```

Comment:
Does that need to be released/free()'d somewhere?

Reply:
I don't think the struct needs to be freed, but I am not sure. I copied this declaration from the sample usage in libinjection repository. The sample code is in C, and in this syntax the struct will be allocated on stack and will be destroyed after the function exits, so there's no need to free after use. Does C++ work the same way?

Reply:
I will try declaring state with "new" and then delete it after use. Will that be more preferable in C++?

Reply:
Quickly looking at libinjection_sqli_init() source code, I don't seem to see a need for any freeing method here. So we;re good.